src/main/java/com/project/cinema/dao/impl/UserDaoImpl.java

```diff
@@ -0,0 +1,47 @@
+package com.project.cinema.dao.impl;
+
+import com.project.cinema.dao.UserDao;
+import com.project.cinema.exceptions.DataProcessingException;
+import com.project.cinema.lib.Dao;
+import com.project.cinema.model.User;
+import com.project.cinema.util.HibernateUtil;
+import java.util.Optional;
+import javax.persistence.criteria.CriteriaBuilder;
+import javax.persistence.criteria.CriteriaQuery;
+import javax.persistence.criteria.Predicate;
+import javax.persistence.criteria.Root;
+import org.hibernate.Session;
+import org.hibernate.Transaction;
+
+@Dao
+public class UserDaoImpl implements UserDao {
+    @Override
+    public User add(User user) {
+        Transaction transaction = null;
+        try (Session session = HibernateUtil.getSessionFactory().openSession()) {
+            transaction = session.beginTransaction();
+            session.save(user);
+            transaction.commit();
+            return user;
+        } catch (Exception e) {
+            if (transaction != null) {
+                transaction.rollback();
+            }
+            throw new DataProcessingException("Can't insert User entity", e);
+        }
+    }
+
+    @Override
+    public Optional<User> findByEmail(String email) {
+        try (Session session = HibernateUtil.getSessionFactory().openSession()) {
```

Comment:
agreed not to use try-with-resources , cause session closing before transaction.rollback(). Use try with finally